src/components/structures/MatrixChat.js

```diff
@@ -1134,26 +1136,65 @@ export default React.createClass({
     /**
      * Called when a new logged in session has started
      */
-    _onLoggedIn: async function() {
+    _onLoggedIn: function() {
         this.setStateForNewView({ view: VIEWS.LOGGED_IN });
-        if (this._is_registered) {
-            this._is_registered = false;
+        if (MatrixClientPeg.currentUserIsJustRegistered()) {
+            MatrixClientPeg.setJustRegisteredUserId(null);
 
             if (this.props.config.welcomeUserId && getCurrentLanguage().startsWith("en")) {
-                const roomId = await createRoom({
-                    dmUserId: this.props.config.welcomeUserId,
-                    // Only view the welcome user if we're NOT looking at a room
-                    andView: !this.state.currentRoomId,
-                });
-                // if successful, return because we're already
-                // viewing the welcomeUserId room
-                // else, if failed, fall through to view_home_page
-                if (roomId) {
-                    return;
+                // We can end up with multiple tabs post-registration where the user
+                // might then end up with a session and we don't want them all making
+                // a chat with the welcome user: try to de-dupe.
+                // We need to wait for the first sync to complete for this to
+                // work though.
+                let waitFor;
+                if (!this.firstSyncComplete) {
+                    waitFor = this.firstSyncPromise.promise;
+                } else {
+                    waitFor = Promise.resolve();
                 }
+                waitFor.then(async () => {
+                    const welcomeUserRooms = DMRoomMap.shared().getDMRoomsForUserId(
+                        this.props.config.welcomeUserId,
+                    );
+                    if (welcomeUserRooms.length === 0) {
+                        const roomId = await createRoom({
+                            dmUserId: this.props.config.welcomeUserId,
+                            // Only view the welcome user if we're NOT looking at a room
+                            andView: !this.state.currentRoomId,
+                        });
+                        // This is a bit of a hack, but since the deduplication relies
+                        // on m.direct being up to date, we need to force a sync
+                        // of the database, otherwise if the user goes to the other
+                        // tab before the next save happens (a few minutes), the
+                        // saved sync will be restored from the db and this code will
+                        // run without the update to m.direct, making another welcome
```

Comment:
firstSync stuff runs after the sync is loaded from the cache (if its there) so it doesn't wait for new data from the server. have commented.

Reply:
I see, and this is preferable to waiting for the catchup sync because you could notice some lag in the riot-chat tab to appear? Just asking because that would seem less hacky to me.

Reply:
Yeah, that would also be an option - I went with doing it this way since then we don't have to wait for a network request on the other side, so don't need to risk leaving the user sitting watching a spinner if the HS/network is down. Agreed this is probably a bit more hacky though.

Reply:
Ok, let's keep it like this, also given that we want this fixed soonish.